### PR TITLE
[Doc Rework] - `Differences regarding the original NGSIv2 spec` Section PR4

### PR DESCRIPTION
**Pull request by mapedraza**
This PR is part of a bigger series #4141

- [x] 15. actionType metadata
   - [x]     - Maps to new section "Differences regarding the  original NGSIv2 spec" section (provisional name)
- [x] 31. Ambiguous subscription status failed not used
   - [x]     - Maps to new section "Differences regarding the  original NGSIv2 spec" section (provisional name)
   - [x]     - status "failed" to be removed from the corresponding table in the subscription datadamodel
- [x] 37. keyValues not supported in POST /v2/op/notify
   - [x]     - To be mentioned in "Differences regarding the  original NGSIv2 spec" section (provisional name)
   - [x]     - Option to be removed from corresponding options table
- [x] 34. Registrations
   - [x]     - The first bullet list maps to new section "Differences regarding the  original NGSIv2 spec" section (provisional name)
   - [x]     - Removed from "Registration Operations" and the registrations datamodel what it is not currently supported
   - [x]     - From "According to NGSIv2 specification..." maps to "Registration Operations" and the registrations datamodel

### CODE REVIEW

File: doc/manuals/orion-api.md [hunk 1]
```diff
@@ -1473,6 +1474,16 @@ The particular alteration type can be got in notifications using the
...
 
+## `actionType` metadata
+
+From original NGSIv2 specification section "Builtin metadata", regarding `actionType` metadata:
+
+> Its value depend on the request operation type: `update` for updates,
+> `append` for creation and `delete` for deletion. Its type is always `Text`.
+
+Current Orion implementation supports `update` and `append`. The `delete` case will be
+supported upon completion of [this issue](https://github.com/telefonicaid/fiware-orion/issues/1494).
```
Comment (mapedraza) [inline doc/manuals/orion-api.md:1485]: It seems it is already completed

File: doc/manuals/orion-api.md [hunk 2]
```diff
@@ -2555,7 +2579,7 @@ A subscription is represented by a JSON object with the following fields:
...
 | [`notification`](#subscriptionnotification) |          | object | An object that describes the notification to send when the subscription is triggered.         |
 | `expires`      | ✓        | ISO8601 | Subscription expiration date in ISO8601 format. Permanent subscriptions must omit this field. |
-| `status`       |          | string | Either `active` (for active subscriptions) or `inactive` (for inactive subscriptions). If this field is not provided at subscription creation time, new subscriptions are created with the `active` status, which can be changed by clients afterwards. For expired subscriptions, this attribute is set to `expired` (no matter if the client updates it to `active`/`inactive`). Also, for subscriptions experiencing problems with notifications, the status is set to `failed`. As soon as the notifications start working again, the status is changed back to `active`. Additionaly, `oneshot` value is available, firing the notification only once whenever the entity is updated after creating the subscription. Once a notification is triggered, the subscription transitions to "status": "inactive". |
```
Comment (mapedraza) [inline doc/manuals/orion-api.md:2083]: Pay attention to this

File: doc/manuals/orion-api.md [hunk 3]
```diff
@@ -2991,7 +3015,7 @@ A context registration is represented by a JSON object with the following fields
...
 | [`dataProvided`](#registrationdataprovided)          |          | object | Object that describes the data provided by this source.                                                                                                                                     |
-| `status`       | ✓        | string | Enumerated field which captures the current status of this registration with the possibles values: [`active`, `inactive`, `expired` or `failed`]. If this field is not provided at registration creation time, new registrations are created with the `active` status, which may be changed by clients afterwards. For expired registrations, this attribute is set to `expired` (no matter if the client updates it to `active`/`inactive`). Also, for registrations experiencing problems with forwarding operations, the status is set to `failed`. As soon as the forwarding operations start working again, the status is changed back to `active`. |
```
Comment (mapedraza) [inline doc/manuals/orion-api.md:2519]: Pay attention to this

File: doc/manuals/orion-api.md [hunk 4]
```diff
@@ -1484,6 +1485,29 @@ From original NGSIv2 specification section "Builtin metadata", regarding `action
 Current Orion implementation supports `update` and `append`. The `delete` case will be
 supported upon completion of [this issue](https://github.com/telefonicaid/fiware-orion/issues/1494).
 
+## Ambiguous subscription status `failed` not used
```
Comment (mapedraza) [inline doc/manuals/orion-api.md:1488]: A future task would need to review subscription payloads examples that include `status` with `failed` value

File: doc/manuals/orion-api.md [hunk 5]
```diff
@@ -1514,6 +1514,22 @@ the subscription failed in its last notification or not (i.e. checking that `fai
 The current Orion implementation doesn't support `keyValues` option in `POST /v2/op/notify` operation described in the 
 original NGSIv2 specification. If you attempt to use it you would get a 400 Bad Request error.
 
+## Registration implementation differences
```
Comment (mapedraza) [inline doc/manuals/orion-api.md:1517]: This new section was created for describing all the differences between Orion and the spec for registrations.

File: doc/manuals/user/ngsiv2_implementation_notes.md [hunk 6]
```diff
@@ -70,20 +70,6 @@ The particular validations that Orion implements on NGSIv2 subscription payloads
...
-* `typePattern` is not implemented.
-* The `expression` field (within `dataProvided`) is not supported. The field is simply
-  ignored. Please see [this issue](https://github.com/telefonicaid/fiware-orion/issues/3107) about it.
-* The `inactive` value for `status` is not supported. I.e., the field is stored/retrieved correctly,
-  but the registration is always active, even when the value is `inactive`. Please see
-  [this issue](https://github.com/telefonicaid/fiware-orion/issues/3108) about it.
-
 According to NGSIv2 specification:
 
 > A NGSIv2 server implementation may implement query or update forwarding to context information sources.
```
Comment (mapedraza) [inline doc/manuals/user/ngsiv2_implementation_notes.md:75]: 34. Registrations 

This point needs to be clarified:

> - [ ] - From "According to NGSIv2 specification..." maps to "Registration Operations" and the registrations datamodel 

Removing the section from `ngsiv2_implementation_notes.md` is pending

Comment (mapedraza) [reply doc/manuals/user/ngsiv2_implementation_notes.md]: Done in https://github.com/telefonicaid/fiware-orion/pull/4183/commits/6e191d86bd9d0dac6b952d1e21d31a2dd180b813

File: doc/manuals/orion-api.md [hunk 7]
```diff
@@ -4196,3 +4150,62 @@ _**Response code**_
 * Successful operation uses 200 OK
 * Errors use a non-2xx and (optionally) an error payload. See subsection on [Error Responses](#error-responses) for
   more details.
+
+# Differences regarding the original NGSIv2 spec
+
+This section contains the topics that, due to implementation decision, the behaviour is differs from the described in the 
+original NGSIv2 specification.
```
Comment (fgalan) [inline doc/manuals/orion-api.md:4157]:
```suggestion
This section contains the topics that, due to implementation decision, differs from the described in the 
original NGSIv2 specification. These differences comes after years of experience with NGSIv2, in two senses:

* Some functionally originally included in NGSIv2 has resulted not to be really useful or needed in real world scenarios. Thus, Orion doesn't implement it. For instance, update registrations operation
* NGSIv2 has some flaws not detected at specification time. For instance, the way in which `status` subscription field was designed.
```

File: doc/manuals/orion-api.md [hunk 8]
```diff
@@ -4196,3 +4150,62 @@ _**Response code**_
...
+
+## `actionType` metadata
+
+From original NGSIv2 specification section "Builtin metadata", regarding `actionType` metadata:
+
+> Its value depend on the request operation type: `update` for updates,
+> `append` for creation and `delete` for deletion. Its type is always `Text`.
+
+Current Orion implementation supports `update` and `append`. The `delete` case will be
+supported upon completion of [this issue](https://github.com/telefonicaid/fiware-orion/issues/1494).
```
Comment (fgalan) [inline doc/manuals/orion-api.md:4167]:
```suggestion
supported upon completion of [this issue](https://github.com/telefonicaid/fiware-orion/issues/2591).
```

I have had a look and it seems the right issue is that one. Please, check.

File: doc/manuals/orion-api.md [hunk 9]
```diff
@@ -3181,7 +3184,7 @@ A subscription is represented by a JSON object with the following fields:
...
+| `status`       |          | string | Either `active` (for active subscriptions) or `inactive` (for inactive subscriptions). If this field is not provided at subscription creation time, new subscriptions are created with the `active` status, which can be changed by clients afterwards. For expired subscriptions, this attribute is set to `expired` (no matter if the client updates it to `active`/`inactive`). Additionaly, `oneshot` value is available, firing the notification only once whenever the entity is updated after creating the subscription. Once a notification is triggered, the subscription transitions to "status": "inactive". |
```
Comment (fgalan) [inline doc/manuals/orion-api.md:3187]:
```suggestion
| `status`       |          | string | Either `active` (for active subscriptions) or `inactive` (for inactive subscriptions). If this field is not provided at subscription creation time, new subscriptions are created with the `active` status, which can be changed by clients afterwards. For expired subscriptions, this attribute is set to `expired` (no matter if the client updates it to `active`/`inactive`). Additionaly, `oneshot` value is available, firing the notification only once whenever the entity is updated after creating the subscription. Once a notification is triggered, the subscription transitions to "status": "inactive". More detail on oneshot subscriptions in the [corresponding section](#oneshot-subscription).|
```

File: doc/manuals/orion-api.md [hunk 10]
```diff
@@ -3631,16 +3635,17 @@ The `provider` field contains the following subfields:
...
 | `attrs`        |          | array | List of attributes to be provided (if not specified, all attributes). |
-| `expression`   |          | object | By means of a filtering expression, allows to express what is the scope of the data provided. Currently only geographical scopes are supported through the following subterms: <ul><li><code>georel</code>: Any of the geographical relationships as specified by the [Geographical queries](#geographical-queries) section. </li><li><code>geometry</code>: Any of the supported geometries as specified by the [Geographical queries](#geographical-queries) section.</li> <li><code>coords</code>: String representation of coordinates as specified by the [Geographical queries](#geographical-queries)section.</li></ul> |
+| `expression`   |          | object | Due to implementation reasons, this filed is ignored. See [Registration implementation differences](#registration-implementation-differences) section |
```
Comment (fgalan) [inline doc/manuals/orion-api.md:3648]:
```suggestion
```

I think is better the API description doesn include "noise" about differences with original NGSIv2 spec. That differences are already detailed in the corresponding section.

File: doc/manuals/orion-api.md [hunk 11]
```diff
@@ -3631,16 +3635,17 @@ The `provider` field contains the following subfields:
...
 |----------------|----------|--------|-----------------------------------------------------------------------------------------------|
-| `entities`     |          | array | A list of objects, each one composed of the following subfields: <ul><li><code>id</code> or <code>idPattern</code>: d or pattern of the affected entities. Both cannot be used at the same time, but one of them must be present.</li><li><code>type</code> or <code>typePattern</code>: Type or pattern of the affected entities. Both cannot be used at the same time. If omitted, it means "any entity type".</li></ul> |
+| `entities`     |          | array | A list of objects, each one composed of the following subfields: <ul><li><code>id</code> or <code>idPattern</code>: d or pattern of the affected entities (only `.*` expression supported at this time(see this [section](#registration-implementation-differences))). Both cannot be used at the same time, but one of them must be present.</li><li><code>type</code> Type of the affected entities. If omitted, it means "any entity type".</li></ul> |
```
Comment (fgalan) [inline doc/manuals/orion-api.md:3646]:
```suggestion
| `entities`     |          | array | A list of objects, each one composed of the following subfields: <ul><li><code>id</code> or <code>idPattern</code>: id or pattern of the affected entities (only the exact expression `.*` is supported). Both cannot be used at the same time, but one of them must be present.</li><li><code>type</code> Type of the affected entities. If omitted, it means "any entity type".</li></ul> |
```

I think is better the API description doesn include "noise" about differences with original NGSIv2 spec. That differences are already detailed in the corresponding section.

File: doc/manuals/orion-api.md [hunk 12]
```diff
@@ -3619,7 +3623,7 @@ A context registration is represented by a JSON object with the following fields
...
+| `status`       | ✓        | string | Enumerated field which captures the current status of this registration with the possibles values: [`active`, `inactive`or `expired`]. If this field is not provided at registration creation time, new registrations are created with the `active` status, which may be changed by clients afterwards. For expired registrations, this attribute is set to `expired` (no matter if the client updates it to `active`/`inactive`). Also, for registrations experiencing problems with forwarding operations, the status is set to `failed`. As soon as the forwarding operations start working again, the status is changed back to `active`. Due to implementation reasons, `inactive` status is not supported. See more infomration in [Registration implementation differences](#registration-implementation-differences) section |
```
Comment (fgalan) [inline doc/manuals/orion-api.md:3626]:
```suggestion
| `status`       | ✓        | string | Enumerated field which captures the current status of this registration with the possibles values: [`active`, `inactive`or `expired`]. If this field is not provided at registration creation time, new registrations are created with the `active` status, which may be changed by clients afterwards. For expired registrations, this attribute is set to `expired` (no matter if the client updates it to `active`/`inactive`). Also, for registrations experiencing problems with forwarding operations, the status is set to `failed`. As soon as the forwarding operations start working again, the status is changed back to `active`. |
```

I think is better the API description doesn include "noise" about differences with original NGSIv2 spec. That differences are already detailed in the corresponding section.

File: doc/manuals/orion-api.md (hunk 3 above)
Comment (fgalan) [reply doc/manuals/orion-api.md:3622]: Let's double check that active, inactive and expired are really working. Not fully sure...

File: doc/manuals/orion-api.md [hunk 13]
```diff
@@ -3595,17 +3598,18 @@ _**Response code**_
...
+* `POST /v2/op/query` for query forwarding
+* `POST /v2/op/update` for update forwarding
 
-* Legacy forwarding (based on NGSIv1 operations)
-* NGSI Context Source Forwarding Specification
+More information on forwarding to context information sources can be found in [this specific document](user/context_providers.md).
 
-Please check the corresponding specification in order to get the details.
+Orion implements an additional field `legacyForwarding` (within `provider`) not included in the NGSIv2
+specification. If the value of `legacyForwarding` is `true` then NGSIv1-based query/update will be used
```
Comment (fgalan) [inline doc/manuals/orion-api.md:3610]:
```suggestion
Orion implements an additional field `legacyForwarding` (within [`provider`](#registrationdataprovided)). If the value of `legacyForwarding` is `true` then NGSIv1-based query/update will be used
```

File: doc/manuals/orion-api.md [hunk 14]
```diff
@@ -3631,16 +3635,17 @@ The `provider` field contains the following subfields:
...
 | `http`         |          | object | It is used to convey parameters for providers that deliver information through the HTTP protocol.(Only protocol supported nowadays). <br>It must contain a subfield named `url` with the URL that serves as the endpoint that offers the providing interface. The endpoint must *not* include the protocol specific part (for instance `/v2/entities`). |
 | `supportedForwardingMode`  |          | string | It is used to convey the forwarding mode supported by this context provider. By default `all`. Allowed values are: <ul><li><code>none</code>: This provider does not support request forwarding.</li><li><code>query</code>: This provider only supports request forwarding to query data.</li><li><code>update</code>: This provider only supports request forwarding to update data.</li><li><code>all</code>: This provider supports both query and update forwarding requests. (Default value).</li></ul> |
+|`legacyForwarding`| ✓      | boolean | If `true`, a NGSIv1-based query/update will be used for forwarding the requests of the registration. |
```
Comment (fgalan) [inline doc/manuals/orion-api.md:3638]:
```suggestion
|`legacyForwarding`| ✓      | boolean | If `true`, a NGSIv1-based query/update will be used for forwarding the requests of the registration. Default (if not included) is `false`. |
```

File: doc/manuals/orion-api.md [hunk 15]
```diff
@@ -4196,3 +4150,62 @@ _**Response code**_
...
+* `PATCH /v2/registration/<id>` is not implemented. Thus, registrations cannot be updated
+  directly. I.e., updates must be done deleting and re-creating the registration. Please
+  see [this issue](https://github.com/telefonicaid/fiware-orion/issues/3007) about this.
+* `idPattern` is supported but only for the exact regular expression `.*`
+* `typePattern` is not implemented.
+* The `expression` field (within `dataProvided`) is not supported. The field is simply
+  ignored. Please see [this issue](https://github.com/telefonicaid/fiware-orion/issues/3107) about it.
+* The `inactive` value for `status` is not supported. I.e., the field is stored/retrieved correctly,
+  but the registration is always active, even when the value is `inactive`. Please see
+  [this issue](https://github.com/telefonicaid/fiware-orion/issues/3108) about it.
```
Comment (fgalan) [inline doc/manuals/orion-api.md:4211]:
```suggestion
  [this issue](https://github.com/telefonicaid/fiware-orion/issues/3108) about it.
* `legacyForwarding` field (within `provider`) to support forwarding in NGSIv1-based query/update format for legacy Context Providers 
```

File: doc/manuals/user/ngsiv2_implementation_notes.md [hunk 16]
```diff
@@ -1,10 +1,7 @@
 # <a name="top"></a>NGSIv2 Implementation Notes
 
 * [Subscription payload validations](#subscription-payload-validations)
-* [`actionType` metadata](#actiontype-metadata)
-* [Ambiguous subscription status `failed` not used](#ambiguous-subscription-status-failed-not-used)
 * [Registrations](#registrations)
```
Comment (fgalan) [inline doc/manuals/user/ngsiv2_implementation_notes.md:4]:
```suggestion
```

File: doc/manuals/orion-api.md (hunk 3 above)
Comment (fgalan) [reply doc/manuals/orion-api.md:3622]: I have realized we say in the differences section:

> The `inactive` value for `status` is not supported

But nothing is told about the `failed` status.

Definitively, we should check how Orion works in this sense to precisely describe the status it supports (and how) for registrations in this document.

File: doc/manuals/orion-api.md (hunk 1 above)
Comment (fgalan) [reply doc/manuals/orion-api.md]: I think was a wrong number. See further comment.

NTC

File: doc/manuals/orion-api.md (hunk 4 above)
Comment (fgalan) [reply doc/manuals/orion-api.md]: Noted in https://github.com/telefonicaid/fiware-orion/pull/4141

File: doc/manuals/orion-api.md (hunk 5 above)
Comment (fgalan) [reply doc/manuals/orion-api.md]: Ok. NTC

File: doc/manuals/orion-api.md [hunk 17]
```diff
@@ -3619,7 +3622,7 @@ A context registration is represented by a JSON object with the following fields
...
+| `status`       | ✓        | string | Enumerated field which captures the current status of this registration with the possibles values: [`active`, `inactive`or `expired`]. If this field is not provided at registration creation time, new registrations are created with the `active` status, which may be changed by clients afterwards. For expired registrations, this attribute is set to `expired` (no matter if the client updates it to `active`/`inactive`). Also, for registrations experiencing problems with forwarding operations, the status is set to `failed`. As soon as the forwarding operations start working again, the status is changed back to `active`. |
```
Comment (fgalan) [inline doc/manuals/orion-api.md:3625]:
```suggestion
| `status`       | ✓        | string | Always `active` in the current implementation. |
```

File: doc/manuals/orion-api.md [hunk 18]
```diff
@@ -4196,3 +4148,66 @@ _**Response code**_
...
+* `PATCH /v2/registration/<id>` is not implemented. Thus, registrations cannot be updated
+  directly. I.e., updates must be done deleting and re-creating the registration. Please
+  see [this issue](https://github.com/telefonicaid/fiware-orion/issues/3007) about this.
+* `idPattern` is supported but only for the exact regular expression `.*`
+* `typePattern` is not implemented.
+* The `expression` field (within `dataProvided`) is not supported. The field is simply
+  ignored. Please see [this issue](https://github.com/telefonicaid/fiware-orion/issues/3107) about it.
+* The `inactive` value for `status` is not supported. I.e., the field is stored/retrieved correctly,
+  but the registration is always active, even when the value is `inactive`. Please see
+  [this issue](https://github.com/telefonicaid/fiware-orion/issues/3108) about it.
```
Comment (fgalan) [inline doc/manuals/orion-api.md:4212]:
```suggestion
* The `inactive` value for `status` is not supported. I.e., the field is stored/retrieved correctly,
  but the registration is always active, even when the value is `inactive`. Please see
  [this issue](https://github.com/telefonicaid/fiware-orion/issues/3108) about it.
* The `expired` value for `status` is not supported. The status is shows as `active` even in the
   registration is actually expired.
```

File: doc/manuals/orion-api.md [hunk 19]
```diff
@@ -3595,17 +3598,17 @@ _**Response code**_
...
-play the role of providers
-of certain subsets (entities, attributes) of the context information space, including those located
-at specific geographical areas.
+play the role of providers of certain subsets (entities, attributes) of the context information space, including those located
+at specific geographical areas. The way in which Orion implements such forwarding is as follows:
 
-A NGSIv2 server implementation may implement query and/or update forwarding to context information sources. In
-particular, some of the following forwarding mechanisms could be implemented (not exhaustive list):
+* `POST /v2/op/query` for query forwarding
+* `POST /v2/op/update` for update forwarding
```
Comment (fgalan) [inline doc/manuals/orion-api.md:3605]:
```suggestion
* [`POST /v2/op/query`](#query-post-v2opquery) for query forwarding
* [`POST /v2/op/update`](#update-post-v2opupdate) for update forwarding
```

File: doc/manuals/orion-api.md [hunk 20]
```diff
@@ -4196,3 +4148,68 @@ _**Response code**_
...
 * Errors use a non-2xx and (optionally) an error payload. See subsection on [Error Responses](#error-responses) for
   more details.
+
+# Differences regarding the original NGSIv2 spec
+
+This section contains the topics that, due to implementation decision, differs from the described in [the 
+original NGSIv2 specification](http://telefonicaid.github.io/fiware-orion/api/v2/stable/). These differences comes after years of experience with NGSIv2, in two senses:
+
+* Some functionally originally included in NGSIv2 has resulted not to be really useful or needed in real world scenarios. Thus, Orion doesn't implement it. For instance, update registrations operation
+* NGSIv2 has some flaws not detected at specification time. For instance, the way in which `status` subscription field was designed.
```
Comment (fgalan) [inline doc/manuals/orion-api.md:4158]:
```suggestion
* NGSIv2 has some flaws not detected at specification time. For instance, the way in which `failed` subscription status was designed.
```

File: doc/manuals/orion-api.md [hunk 21]
```diff
@@ -4196,3 +4148,68 @@ _**Response code**_
 * Successful operation uses 200 OK
 * Errors use a non-2xx and (optionally) an error payload. See subsection on [Error Responses](#error-responses) for
   more details.
+
+# Differences regarding the original NGSIv2 spec
+
+This section contains the topics that, due to implementation decision, differs from the described in [the 
+original NGSIv2 specification](http://telefonicaid.github.io/fiware-orion/api/v2/stable/). These differences comes after years of experience with NGSIv2, in two senses:
```
Comment (fgalan) [inline doc/manuals/orion-api.md:4155]:
```suggestion
original NGSIv2 specification](http://telefonicaid.github.io/fiware-orion/api/v2/stable/). These differences come after years of experience with NGSIv2, in two senses:
```

**Review by fgalan — APPROVED** [review]
LGTM

**Pull request merged**